### ISSUE TITLE
Elaborate on Step organization

### ISSUE THREAD

**Issue by mlvandijk** [cucumber/docs#135]
Step organization could use some more elaboration, especially on reuse of helper methods: https://docs.cucumber.io/gherkin/step-organization/

**Comment by mlvandijk** [comment]
From Slack: 
NaviHan   [Apr 6th at 10:53]
Is there any good documentation available to understand the use of helper methods in Cucumber?
21 replies

gazler   [2 days ago]
You can read a brief explanation with an example here :
https://docs.cucumber.io/gherkin/step-organization/

gazler   [2 days ago]
But in general, helper methods are not cucumber specific. Any method that you write to separate your glue code from implementation logic or to reuse in multiple steps (thus reducing redundant steps) can be considered a helper method

NaviHan   [2 days ago]
I have read this. But no examples

NaviHan   [2 days ago]
Is this explained in any books?

gazler   [2 days ago]
The *Grouping step definitions* section shows an example on how to use a factory method to open different pages using one method

gazler   [2 days ago]
You can read up on factory methods if you need to know how a factory pattern is used to achieve something similar. There's a link provided in the same section for that as well.

NaviHan   [2 days ago]
Thanks @gazler this is still very abstract. Im looking the implementation as well with examples

gazler   [2 days ago]
Welcome :slightly_smiling_face: Explaining an implementation of the factory pattern won't be under the scope of cucumber any longer, so it is not included. Like I suggested, you can read up on patterns to understand how the inner logic would work for the given example, it should help :slightly_smiling_face:

NaviHan   [2 days ago]
Sure. Thanks :slightly_smiling_face:

Cat   [2 days ago]
I also highly recommend patterns like the page object model for UI automation for building scalable infrastructure for UI testing + related helpers

js   [2 days ago]
The simplest way I can explain helper methods is: methods that do things for you. (rather than the step definitions)

js   [2 days ago]
Think you need to have a step definition format some json for you? Nope, a helper method should do it.

Cat   [2 days ago]
^^ This precisely.

Cat   [2 days ago]
Rather than essentially rewriting the same function that is very slightly modified as a stepDefinition, it scales better if you write a more abstract helper, then leverage that for your stepDefinition.

mlvandijk   [1 day ago]
Also, any examples would depend on the context.

mlvandijk   [1 day ago]
The Cucumber (for Java) book might have something?

mlvandijk   [1 day ago]
Uploading a pic of how I tried to explain to devs on my team (trying anyway)

mlvandijk   [1 day ago]
I like to group my stepdefs according to domain objects and to keep them very thin. Ie only the "glue" between step and code. Where most of my code is in the helper methods. Which I can call from any stepdef.

mlvandijk   [1 day ago]
The helpers will be either mocks (for services outside of our scope) or interactions with the appliction ( Api calls to endpoint and/ page objects for the UI).

mlvandijk   [1 day ago]
Ah the pic is in the main thread. Hope you can read it

mlvandijk   [12 hours ago]
Been meaning to add this to the docs... (edited)

![20190406_205950](https://user-images.githubusercontent.com/20927578/55747982-ca986280-5a3d-11e9-89b3-040d6ae391ff.jpg)

**Comment by zutshiy** [comment]
Adding further info relevant to this issue from the above thread :


> Andrew Premdas   [3 hours ago]
> The main thing Cucumber needs to document is how to get a step definition to make a call from a step definition to a helper method. For ruby this is very easy :
> In your step definition make your call to a helper method .e.g
```
When "I foo" do
    foo
end
```
> this will call a method in the Cucumber World. To add the method to the Cucumber World you need to
> 1 - define a method inside a module
> 2 - add your module to the Cucumber World
> 
> e.g.
 this is a file in features/support or features/step_definitions
```
module MyStepHelper
  def foo
  end
end
World MyStepHelper
```
> 
> Andrew Premdas   [1 hour ago]
> I think you need to have a separate new section at the level of step organisation for this, perhaps ‘Steps and Helper Methods’ and then have a subsection inside that e.g. ‘Defining a helper method and calling it from a step definition’ which is language specific. Other sections in here might include
> 
> Steps as a translation layer  # generic: explanation of the pattern of each step being just a call
> Why helper methods            # generic: explain about pushing the HOW down and how code is designed for complexity and cucumber isn’t.
>But really the key thing is to get simple language explanations for each language to fill  ‘Defining a helper method and calling it from a step definition’

**Comment by zutshiy** [comment]
![HelperMethods](https://user-images.githubusercontent.com/34692418/55966270-b51d7580-5c95-11e9-8132-9bad80e6a64b.png)

**Comment by mxygem** [comment]
NICE work @gazler22! This is really looking great. I've got some additional requests based on this iteration.

1. Remove the dots below the list of feature files and center those objects in the rectangle?
2. Straighten the arrow body where it connects to the helper methods object.
3. Move the description of the helper methods object to outside and below to match the other objects
4. Bring the size of the helper methods object to be more similar to the others.
5. Perhaps remove the lightened green from the feature files object to try to keep things similar to the others, too?

**Comment by zutshiy** [comment]
![HelperMethods](https://user-images.githubusercontent.com/34692418/55979862-c1fb9280-5cb0-11e9-8e32-bd381009be85.png)


Updated.

**Comment by mxygem** [comment]
NICE! One final tweak (I think? hah) can we get the descriptions to be even with one another vertically?

**Comment by zutshiy** [comment]
Alright! Two variants! Geez, this diagram business is getting me excited, lol.

![Helper Methods 2](https://user-images.githubusercontent.com/34692418/55982613-e9555e00-5cb6-11e9-9bf7-7ee016f68923.png)
![Helper Methods 3](https://user-images.githubusercontent.com/34692418/55982614-e9555e00-5cb6-11e9-8b54-8c98d5137555.png)

**Comment by mxygem** [comment]
Hah! Thank you for being flexible. :) I like the first one more, personally.

**Comment by zutshiy** [comment]
I like both. I just drew the second one because the description seemed a bit too disconnected from it's  image after the alignment.

But first one, it is!

**Comment by mlvandijk** [comment]
Please see #435

**Comment by mlvandijk** [comment]
Fixed by #435
